### ISSUE TITLE
chapter 11 Reusing python code 

### ISSUE THREAD

**Issue by nitml** [ageron/handson-ml#272]
My question might be foolish.In reusing python code section in jupyter notebook  below the line "Alternatively, if you have access to the Python code that built the original graph, you can use it instead of `import_meta_graph()`" code written was:

```python
reset_graph()

n_inputs = 28 * 28  # MNIST
n_hidden1 = 300
n_hidden2 = 50
n_hidden3 = 50
n_hidden4 = 50
n_outputs = 10

X = tf.placeholder(tf.float32, shape=(None, n_inputs), name="X")
y = tf.placeholder(tf.int32, shape=(None), name="y")

with tf.name_scope("dnn"):
    hidden1 = tf.layers.dense(X, n_hidden1, activation=tf.nn.relu, name="hidden1")
    hidden2 = tf.layers.dense(hidden1, n_hidden2, activation=tf.nn.relu, name="hidden2")
    hidden3 = tf.layers.dense(hidden2, n_hidden3, activation=tf.nn.relu, name="hidden3")
    hidden4 = tf.layers.dense(hidden3, n_hidden4, activation=tf.nn.relu, name="hidden4")
    hidden5 = tf.layers.dense(hidden4, n_hidden5, activation=tf.nn.relu, name="hidden5")
    logits = tf.layers.dense(hidden5, n_outputs, name="outputs")

with tf.name_scope("loss"):
    xentropy = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=y, logits=logits)
    loss = tf.reduce_mean(xentropy, name="loss")

with tf.name_scope("eval"):
    correct = tf.nn.in_top_k(logits, y, 1)
    accuracy = tf.reduce_mean(tf.cast(correct, tf.float32), name="accuracy")

learning_rate = 0.01
threshold = 1.0

optimizer = tf.train.GradientDescentOptimizer(learning_rate)
grads_and_vars = optimizer.compute_gradients(loss)
capped_gvs = [(tf.clip_by_value(grad, -threshold, threshold), var)
              for grad, var in grads_and_vars]
training_op = optimizer.apply_gradients(capped_gvs)

init = tf.global_variables_initializer()
saver = tf.train.Saver()

with tf.Session() as sess:
    saver.restore(sess, "./my_model_final.ckpt")

    for epoch in range(n_epochs):
        for X_batch, y_batch in shuffle_batch(X_train, y_train, batch_size):
            sess.run(training_op, feed_dict={X: X_batch, y: y_batch})
        accuracy_val = accuracy.eval(feed_dict={X: X_valid, y: y_valid})
        print(epoch, "Validation accuracy:", accuracy_val)

    save_path = saver.save(sess, "./my_new_model_final.ckpt")
```

MY QUESTION is why `saver.restore(sess, "./my_model_final.ckpt")` was used in place of `sess.run(init)` ? As in this code we are not using any other model so why we are using `saver.restore()` ?

**Comment by ageron** [comment]
Hi @nitml ,
Thanks for your question.  When you restore a pretrained model, you need two things:
1) the graph structure
2) the variable values

To restore the graph structure, there are two options:
1a) use `import_meta_graph()`
1b) manually build the graph structure (that's only if you have the exact code that was used to generate the graph whose variable values were saved previously).

In this section, I wanted to show option 1b.  From the beginning up to the line `saver = tf.train.Saver()`, the code is exactly the same as earlier, when we created the model, trained it and saved it. In other words we create the exact same graph.  But this only restores the graph structure, not the variable values.  We now want to go to step 2 and restore the variable values.  For this we create a session and we use the saver to restore the variable values that were previously saved to disk.

After that we can use the model any way we want, for example we can continue training and evaluate the model on the validation set, then save the improved model.

If we used `sess.run(init)` instead of `saver.restore()`, we would not be restoring the model saved earlier, we would just we initializing a new model.  It would start training from scratch instead of continuing where we left off earlier.

Hope this makes sense,
Aurélien

**Comment by nitml** [comment]
thanks for your time and for explaining it ....
